python/lsst/validate/drp/matchreduce.py

```diff
@@ -203,30 +214,56 @@ def _loadAndMatchCatalogs(self, repo, dataIds, matchRadius):
+                except (FitsError, dafPersist.NoResults) as e:
+                    print(e)
+                    print("Could not open updated WCS for ", vId)
+                    print("Skipping %s " % repr(vId))
+                    continue
+            else:
+                try:
+                    calexpMetadata = butler.get("calexp_md", vId)
+                except (FitsError, dafPersist.NoResults) as e:
+                    print(e)
+                    print("Could not open calibrated image file for ", vId)
+                    print("Skipping %s " % repr(vId))
+                    continue
+                except TypeError as te:
+                    # DECam images that haven't been properly reformatted
+                    # can trigger a TypeError because of a residual FITS header
+                    # LTV2 which is a float instead of the expected integer.
+                    # This generates an error of the form:
+                    #
+                    # lsst::pex::exceptions::TypeError: 'LTV2 has mismatched type'
+                    #
+                    # See, e.g., DM-2957 for details.
+                    print(te)
+                    print("Calibration image header information malformed.")
+                    print("Skipping %s " % repr(vId))
+                    continue
+
+                calib = afwImage.Calib(calexpMetadata)
+
+            # We don't want to put this above the first "if useJointCal block"
+            # because we need to use it to quickly catch data IDs with no
+            # usable outputs.
             try:
-                calexpMetadata = butler.get("calexp_md", vId, immediate=True)
-            except (FitsError, dafPersist.NoResults) as e:
-                print(e)
-                print("Could not open calibrated image file for ", vId)
-                print("Skipping %s " % repr(vId))
-                continue
-            except TypeError as te:
-                # DECam images that haven't been properly reformatted
-                # can trigger a TypeError because of a residual FITS header
-                # LTV2 which is a float instead of the expected integer.
-                # This generates an error of the form:
-                #
-                # lsst::pex::exceptions::TypeError: 'LTV2 has mismatched type'
-                #
-                # See, e.g., DM-2957 for details.
-                print(te)
-                print("Calibration image header information malformed.")
-                print("Skipping %s " % repr(vId))
-                continue
-
-            calib = afwImage.Calib(calexpMetadata)
-
-            oldSrc = butler.get('src', vId, immediate=True)
+                # HSC supports these flags, which dramatically improve I/O
+                # performance; support for other cameras is DM-6927.
+                oldSrc = butler.get('src', vId, flags=SOURCE_IO_NO_FOOTPRINTS)
+            except:
+                oldSrc = butler.get('src', vId)
```

Comment:
Big improvement in I/O time, mostly.  It's a big relative change in memory consumption too, but I don't notice that as much.